player_test.py

```diff
@@ -473,5 +473,223 @@ def test_ant_moves_from_start_to_anthill_if_none_of_its_food_in_trail(self):
+
+  def test_ant_is_in_trail_but_has_only_empty_spaces_behind_returns_only_front(self):
+  # test 62
+    mario = Player("mario")
+    ant = "green"
+    trail = [None, "pepper", "apple", "cheese"]
+    ant_positions = {"green": 1}
+    expected_allowed_choices = ["front"]
+    self.assertEqual(mario.define_allowed_choices_direction(ant, trail, ant_positions), expected_allowed_choices)
+    self.assertNotIn("back", mario.define_allowed_choices_direction(ant, trail, ant_positions))
+
+  def test_ant_is_in_trail_but_has_only_empty_spaces_in_front_returns_only_back(self):
+  # test 63
+    mario = Player("mario")
+    ant = "yellow"
+    trail = ["pepper", "apple", "cheese", None]
+    ant_positions = {"yellow": 2}
+    expected_allowed_choices = ["back"]
+    self.assertEqual(mario.define_allowed_choices_direction(ant, trail, ant_positions), expected_allowed_choices)
+    self.assertNotIn("front", mario.define_allowed_choices_direction(ant, trail, ant_positions))
+
+  def test_presence_of_other_ants_invalidates_food_so_direction_back_not_included(self):
+  # test 64
+    mario = Player("mario")
+    ant = "green"
+    trail = [None, "grapes", None, "pepper", "apple", "cheese"]
+    ant_positions = {"green": 3, "purple": 1}
+    expected_allowed_choices = ["front"]
+    self.assertEqual(mario.define_allowed_choices_direction(ant, trail, ant_positions), expected_allowed_choices)
+    self.assertNotIn("back", mario.define_allowed_choices_direction(ant, trail, ant_positions))
+
+  def test_presence_of_other_ants_invalidates_food_so_direction_front_not_included(self):
+  # test 65
+    mario = Player("mario")
+    ant = "yellow"
+    trail = ["pepper", "apple", "cheese", None, "grapes", None]
+    ant_positions = {"yellow": 2, "purple": 4}
+    expected_allowed_choices = ["back"]
+    self.assertEqual(mario.define_allowed_choices_direction(ant, trail, ant_positions), expected_allowed_choices)
+    self.assertNotIn("front", mario.define_allowed_choices_direction(ant, trail, ant_positions))
+
+class TakeTurnTest(unittest.TestCase):
+  def test_single_ant_moves_to_centre_of_three_element_trail_picks_front(self):
+  # test 66
+    # Given
+    anthill = [None]
+    mario = Player("mario")
+    trail = ["pepper", "apple", "cheese"]
+    ant_positions = {"red": None}
+    input_patcher = mock.patch('builtins.input', side_effect = ["red", "front"])
+    input_mock = input_patcher.start()
+    # When
+    (actual_new_trail, actual_new_ant_positions, actual_new_anthill) = \
+      mario.take_turn(trail, ant_positions, anthill)
+    # Then
+    expected_new_hand = {"cheese": 1}
+    expected_new_trail = ["pepper", "apple", None]
+    expected_new_ant_positions = {"red": 1}
+    expected_new_anthill = [None]
+    self.assertEqual(actual_new_trail, expected_new_trail)
+    self.assertEqual(actual_new_ant_positions, expected_new_ant_positions)
+    self.assertEqual(actual_new_anthill, expected_new_anthill)
+    self.assertEqual(mario.hand, expected_new_hand)
```

Comment:
Following our earlier conversation about pure functions, I suggest (optionally) checking that the original trail, ant_positions, and anthill have not been changed.